dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java

```diff
@@ -927,214 +938,361 @@ private String getContentWDL(File tempMainDescriptor, Type type) {
-                String keyStep = stepMap.toString();
-                Object valueStep = stepsMap.get(stepMap);
-                Map<String, Object> stepMapValue = new HashMap<>();
-                stepMapValue.put(keyStep,valueStep);
-                steps.add(stepMapValue);
+     * Given a list of CWL requirements, will return the DockerPull information if present.
+     * If not will return the current docker path (currentDefault)
+     * @param requirements
+     * @param currentDefault
+         * @return
+         */
+    private String getDockerRequirement(List<Object> requirements, String currentDefault) {
+        if (requirements != null) {
+            for (Object requirement : requirements) {
+                // TODO : currently casting to map, but should use CWL Avro classes
+                //            if (requirement instanceof DockerRequirement) {
+                //                if (((DockerRequirement) requirement).getDockerPull() != null) {
+                //                    LOG.info(((DockerRequirement) requirement).getDockerPull().toString());
+                //                    dockerPath = ((DockerRequirement) requirement).getDockerPull().toString();
+                //                    break;
+                //                }
+                //            }
+                if (((Map) requirement).get("class").equals("DockerRequirement") && ((Map) requirement).get("dockerPull") != null) {
+                    return ((Map) requirement).get("dockerPull").toString();
+                }
             }
-        }else {
-            steps = (ArrayList<Map <String, Object>>) entry.getValue();
         }
-        return steps;
+
+        return currentDefault;
     }
 
     /**
-     * This method will check for the docker environment of the workflow by looking for "hints" and "requirements"
-     * and make it as default docker environment (if available, else return "")
-     * @param section
-     * @param entry
-     * @param defaultDockerEnv
-     * @return defaultDockerEnv
-     */
-    private String getDefaultDockerEnv(String section,Map.Entry<String, Object> entry, String defaultDockerEnv){
-        if (section.equals("hints")) {
-            //docker requirement of the workflow
-            ArrayList<Map<String, Object>> requirements = (ArrayList<Map<String, Object>>) entry.getValue();
-            for (Map <String, Object> requirement : requirements) {
-                if (requirement.get("class").equals("DockerRequirement")) {
-                    defaultDockerEnv = requirement.get("dockerPull").toString();
+     * Given a list of CWL hints, will return the DockerPull information if present.
+     * If not will return the current docker path (currentDefault)
+     * @param hints
+     * @param gsonWorkflow
+     * @param currentDefault
+         * @return
+         */
+    private String getDockerHint(List<Any> hints, Gson gsonWorkflow, String currentDefault) {
```

Comment:
This is a getting a bit long, it probably makes sense to move this into a dedicated graph generator class, also helps out with re-use in a different project if it comes to that.